feat: add fordefi signer to aa-signers

gilmeir-arnac opened the pull request:
# Pull Request Checklist

- [x] Did you add new tests and confirm existing tests pass? (`yarn test`)
- [x] Did you update relevant docs? (docs are found in the `site` folder, and guidelines for updating/adding docs can be found in the [contribution guide](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md))
- [x] Do your commits follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [x] Does your PR title also follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [x] If you have a breaking change, is it [correctly reflected in your commit message](https://www.conventionalcommits.org/en/v1.0.0/#examples)? (e.g. `feat!: breaking change`)
- [x] Did you run lint (`yarn lint:check`) and fix any issues? (`yarn lint:write`)
- [x] Did you follow the [contribution guidelines](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md)?

<!-- start pr-codex -->

---

## PR-Codex overview
This PR adds the `Fordefi` signer and related methods, updates dependencies, and includes documentation for the `FordefiSigner` in the project.

### Detailed summary
- Added `FordefiSigner` and related methods
- Updated dependencies with `@fordefi/web3-provider`
- Included documentation for `FordefiSigner` in `aa-signers` package

> The following files were skipped due to too many changes: `site/packages/aa-signers/fordefi/constructor.md`, `site/packages/aa-signers/fordefi/introduction.md`, `packages/signers/src/fordefi/signer.ts`, `site/signers/guides/fordefi.md`, `packages/signers/src/fordefi/__tests__/signer.test.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

<!-- end pr-codex -->

gilmeir-arnac commented:
Hey @moldy530 this PR comes instead #657 targeting `main` this time. Could you please approve the workflow run? 🙏

moldy530 commented on yarn.lock:
```diff
@@ -4156,6 +4156,13 @@
+  resolved "https://registry.yarnpkg.com/@fordefi/web3-provider/-/web3-provider-0.2.0.tgz#c48da777410081e29e6ccdae6e40c8dc1558f395"
+  integrity sha512-7zBzg/F8fJgspmjeKbI3QjNKLYmycCtWOuPf225Ye0O+8zRtQr6PgaCh2tDhmFMa3gmAJ+BAhuPqCyYyNw4Y+w==
+  dependencies:
+    viem "^2.8.6"
```
oh this might cause some issues since we need to pin our SDK to `2.8.6` having `^2.8.6` as a dependency will have yarn (as you can see below it maps `^2.8.6` to `2.9.32`) pull in some other version.

gilmeir-arnac replied:
removed

gilmeir-arnac commented:
@moldy530 Thanks for reviewing, I fixed the comment.

The [lint job failed](https://github.com/alchemyplatform/aa-sdk/actions/runs/9084400426/job/24966855436#step:2:16) with `Error: Input required and not supplied: token`. It seems that the secret `REVIEWDOG_GITHUB_API_TOKEN` is [required](https://github.com/alchemyplatform/aa-sdk/blob/main/.github/workflows/on-pull-request.yml#L37), would I need to set up a github personal access token myself in my fork's settings?

moldy530 commented:
oh right.... I'm not entirely sure what the fix is for that. I'll have to figure it out. in the meantime: 
1. run `yarn lint:write` locally
2. for any public methods you've written can you add jsdoc comments? They should be in the format of :
```
/**
 * {description}
 * 
 * @param {paramName} {description}
 * @returns {description}
 */
```

gilmeir-arnac commented:
@moldy530
1. added jsdoc to public methods. 
2. `yarn lint:write` passes locally.
3. cleaned up redundant type aliases to `void` for readability

got an idea how to proceed with the `token` error in github actions? secrets on gh actions may not be propagating to workflow executions on forks like here.. I see that some other PRs from other third parties were created by devs that are collaborators in this repo, is that your process?

gilmeir-arnac commented:
Hey guys
Would really appreciate your assistance pushing this PR forward 🙏 
@moldy530 @dancoombs @rthomare @mokok123 @dphilipson 
Thanks

moldy530 commented:
> got an idea how to proceed with the token error in github actions

I think it's ok here, if your tests are running locally and you've linted then it's fine. Most likely you will need to set a GH action secret as a PAT that will allow the action to make comments to your PRs.


This LGTM overall, but something I want to call out:
On our roadmap for v4 is to actually deprecate aa-signers and aa-accounts because (as you can see) maintaining 3rd party integrations is not super scalable for us. Most libraries invert this control where developers needing to provide compatibility with specific libraries will publish and maintain their own packages. A good example is `@turnkey/viem` which is maintained by the turnkey team to provide viem compatible interfaces for their core sdk.

I will approve and get this landed, but v4 will drop this support. We aim to work on v4 next quarter which will introduce some other breaking changes. So just a heads up!
